service/worker/sysworkflow/system_workflow.go

```diff
@@ -22,103 +22,414 @@ package sysworkflow
+	for requestsHandled < signalsUntilContinueAsNew {
+		var request ArchiveRequest
+		if more := ch.Receive(ctx, &request); !more {
 			break
 		}
-		selectSystemTask(scope, signal, ctx, logger)
+		metricsClient.IncCounter(metrics.SystemWorkflowScope, metrics.SysWorkerReceivedSignal)
+		requestsHandled++
+		workQueue.Send(ctx, request)
 	}
 
+	// step 4: drain signal channel to get next run's carryover
+	var co []ArchiveRequest
 	for {
-		var signal signal
-		if ok := ch.ReceiveAsync(&signal); !ok {
+		var request ArchiveRequest
+		if ok := ch.ReceiveAsync(&request); !ok {
 			break
 		}
-		selectSystemTask(scope, signal, ctx, logger)
-		signalsHandled++
+		metricsClient.IncCounter(metrics.SystemWorkflowScope, metrics.SysWorkerReceivedSignal)
+		co = append(co, request)
 	}
 
-	logger.Info("completed current set of iterations, continuing as new",
-		zap.Int(logging.TagIterationsUntilContinueAsNew, signalsHandled))
+	// step 5: wait for all in progress work to finish
+	for i := 0; i < requestsHandled; i++ {
+		finishedWorkQueue.Receive(ctx, nil)
+	}
 
-	ctx = workflow.WithExecutionStartToCloseTimeout(ctx, WorkflowStartToCloseTimeout)
-	ctx = workflow.WithWorkflowTaskStartToCloseTimeout(ctx, DecisionTaskStartToCloseTimeout)
-	return workflow.NewContinueAsNewError(ctx, SystemWorkflowFnName)
+	// step 6: schedule new run
+	ctx = workflow.WithExecutionStartToCloseTimeout(ctx, workflowStartToCloseTimeout)
+	ctx = workflow.WithWorkflowTaskStartToCloseTimeout(ctx, decisionTaskStartToCloseTimeout)
+	metricsClient.IncCounter(metrics.SystemWorkflowScope, metrics.SysWorkerContinueAsNew)
+	sw.Stop()
+	logger.WithFields(bark.Fields{
+		logging.TagNumberOfSignalsUntilContinueAsNew: requestsHandled,
+	}).Info("system workflow is continuing as new")
+	return workflow.NewContinueAsNewError(ctx, archiveSystemWorkflowFnName, co)
 }
 
-func selectSystemTask(scope tally.Scope, signal signal, ctx workflow.Context, logger *zap.Logger) {
-	scope.Counter(HandledSignalCount).Inc(1)
-
-	ao := workflow.ActivityOptions{
+func handleRequest(request ArchiveRequest, ctx workflow.Context, logger bark.Logger, metricsClient metrics.Client) {
+	uploadActOpts := workflow.ActivityOptions{
+		ScheduleToStartTimeout: time.Minute,
+		StartToCloseTimeout:    5 * time.Minute,
+		HeartbeatTimeout:       heartbeatTimeout,
+		RetryPolicy: &cadence.RetryPolicy{
+			InitialInterval:    time.Second,
+			BackoffCoefficient: 2.0,
+			MaximumInterval:    time.Minute,
+			ExpirationInterval: time.Hour * 24 * 30,
+			NonRetriableErrorReasons: []string{
+				errArchivalUploadActivityGetDomainStr,
+				errArchivalUploadActivityNextBlobStr,
+				errArchivalUploadActivityConstructKeyStr,
+				errArchivalUploadActivityBlobExistsStr,
+				errArchivalUploadActivityMarshalBlobStr,
+				errArchivalUploadActivityConvertHeaderToTagsStr,
+				errArchivalUploadActivityWrapBlobStr,
+				errArchivalUploadActivityUploadBlobStr,
+			},
+		},
+	}
+	if err := workflow.ExecuteActivity(
```

Comment:
Is there a way to interrupt long running activities if it is time to continueAsNew and one or more activities are continuously in retry.

Reply:
Will address in text CR.